packages/loopring_v3/contracts/converters/BaseConverter.sol

```diff
@@ -0,0 +1,202 @@
+abstract contract BaseConverter is LPERC20, Claimable, Drainable
+{
+    using AddressUtil       for address;
+    using ERC20SafeTransfer for address;
+    using MathUint          for uint;
+
+    event ConversionSuccess (uint amountIn, uint amountOut);
+    event ConversionFailed  (string reason);
+
+    IExchangeV3        public  immutable exchange;
+    IDepositContract   public  immutable depositContract;
+
+    bool               public  initialized;
+
+    address            public  tokenIn;
+    address            public  tokenOut;
+
+    bool               public  failed;
+
+    modifier onlyFromExchangeOwner()
+    {
+        require(msg.sender == exchange.owner(), "UNAUTHORIZED");
+        _;
+    }
+
+    constructor(
+        IExchangeV3 _exchange
+        )
+    {
+        exchange = _exchange;
+        depositContract = _exchange.getDepositContract();
+    }
+
+    function initialize(
+        string memory _name,
+        string memory _symbol,
+        uint8         _decimals,
+        address       _tokenIn,
+        address       _tokenOut
+        )
+        external
+    {
+        require(!initialized, "ALREADY_INITIALIZED");
+        initializeToken(_name, _symbol, _decimals);
+
+        tokenIn = _tokenIn;
+        tokenOut = _tokenOut;
+
+        initialized = true;
+    }
+
+    function deposit(
+        uint96          amountIn,
+        uint96          minAmountOut,
+        bytes  calldata customData
+        )
+        external
+        payable
+        onlyFromExchangeOwner
+    {
+        require(totalSupply == 0);
+
+        // Converter specific logic, which can fail
+        try BaseConverter(this).convertExternal(amountIn, minAmountOut, customData)
+            returns (uint amountOut) {
+            failed = false;
+            emit ConversionSuccess(amountIn, amountOut);
+        } catch Error(string memory reason) {
+            failed = true;
+            emit ConversionFailed(reason);
+        } catch {
+            failed = true;
+            emit ConversionFailed("unknown");
+        }
+
+        // Mint pool tokens representing each user's share in the pool, with 1:1 ratio
+        _mint(address(this), amountIn);
+
+        // Repay the flash mint used to give user's their share on L2
+        _repay(address(this), amountIn);
+    }
+
+    function withdraw(
+        address to,
+        uint96  poolAmount,
+        uint96  repayAmount
+        )
+        public
+    {
+        // Token to withdraw
+        address token = failed ? tokenIn : tokenOut;
+
+        // Current balance in the contract
+        uint balance = 0;
+        if (token == address(0)) {
+            balance = address(this).balance;
+        } else {
+            balance = ERC20(token).balanceOf(address(this));
+        }
+
+        // Share to withdraw
+        uint amount = balance.mul(poolAmount) / totalSupply;
+
+        // Burn pool tokens
+        _burn(msg.sender, poolAmount);
```

Comment:
This pool token can only be owned by the DepositContract, so only the DepositContract can call this function?  I'm having a hard time understand who are supposed to be the msg.sender.

Reply:
I should have explained this somewhere a bit better, the test in `testConverter.ts` makes it more clear what happens but I'll describe it here.

Phase 1:
- First flash mint the necessary vault tokens
- Do the phase 1 trades to the vault token
- Withdraw tokens collected to the convert address (!). This is because the L1 swap will be done in the converter, so it's cheaper to just transfer the tokens there directly. There's also no risk someone else being able to use those funds because the withdrawal transfer + converter deposit happens in the same transaction.
- Than we temporarily mint the vault tokens to the convert address because that's also the address that will repay the flash loan
- Use those minted tokens to repay the flash loan

Phase 2:
- First flash mint the necessary tokens people want to actually buy
- Do the phase 2 trades to the token the user wanted to buy
- Withdraw tokens collected to the operator contract address (!). The operator contract will do a callback to will batch withdraw from the vault by burning all vault tokens together.
- The tokens unlocked by burning the vault tokens are now used to repay the flash loan

Reply:
I think I also figured it out last night. I'll explain the whole process to the relayer team for them to implement. I will forward questions that I cannot answer :)